# ICP04 Assignment

## Aravind Dudam

## U04440266

## Library Import¶

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier 
from matplotlib import pyplot as plt
import numpy as np
from sklearn.metrics import confusion_matrix
from sklearn import preprocessing
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from scipy import stats

np.random.seed(1)

##  Data Load

In [2]:
X_train = pd.read_csv('E:/DSP/Week 04/airbnb_train_X_price_gte_150.csv')
y_train = pd.read_csv('E:/DSP/Week 04/airbnb_train_y_price_gte_150.csv')
X_test = pd.read_csv('E:/DSP/Week 04/airbnb_test_X_price_gte_150.csv')
y_test = pd.read_csv('E:/DSP/Week 04/airbnb_test_y_price_gte_150.csv') 

In [12]:
X_train

host_is_superhost  host_identity_verified  latitude  longitude  \
0                     0                       0  0.884939   1.012890   
1                     0                       1 -1.808025   0.861706   
2                     0                       1  0.215255  -0.095632   
3                     0                       0 -1.857605  -1.058074   
4                     0                       0 -1.199298   1.125934   
...                 ...                     ...       ...        ...   
2483                  0                       1 -1.482015   0.675233   
2484                  0                       1 -0.023296   0.126650   
2485                  0                       1  0.177577   0.691157   
2486                  0                       1 -1.457699  -1.419131   
2487                  0                       1 -0.123707   0.262572   

      room_type  accommodates  bathrooms  bedrooms      beds  bed_type  ...  \
0             0      0.539411  -0.441906  0.999675  0.375655         4  ...   
1             0      0.539411  -0.441906  2.325470  1.333267         4  ...   
2             0     -0.017689  -0.441906 -0.326120  0.375655         4  ...   
3             1     -0.574789  -0.441906 -0.326120 -0.581957         4  ...   
4             0     -0.574789  -0.441906 -0.326120 -0.581957         4  ...   
...         ...           ...        ...       ...       ...       ...  ...   
2483          1      0.539411   0.564007 -0.326120 -0.581957         4  ...   
2484          1     -1.131889  -0.441906 -0.326120 -0.581957         4  ...   
2485          0     -0.574789  -0.441906 -0.326120 -0.581957         4  ...   
2486          1     -1.131889   0.564007 -0.326120 -0.581957         4  ...   
2487          1     -0.574789   0.564007 -0.326120 -0.581957         4  ...   

      property_type_Condominium  property_type_Dorm  \
0                             0                   0   
1                             1                   0   
2                             0                   0   
3                             1                   0   
4                             0                   0   
...                         ...                 ...   
2483                          0                   0   
2484                          0                   0   
2485                          1                   0   
2486                          0                   0   
2487                          1                   0   

      property_type_Entire Floor  property_type_Guesthouse  \
0                              0                         0   
1                              0                         0   
2                              0                         0   
3                              0                         0   
4                              0                         0   
...                          ...                       ...   
2483                           0                         0   
2484                           0                         0   
2485                           0                         0   
2486                           0                         0   
2487                           0                         0   

      property_type_House  property_type_Loft  property_type_Other  \
0                       0                   0                    0   
1                       0                   0                    0   
2                       0                   0                    0   
3                       0                   0                    0   
4                       1                   0                    0   
...                   ...                 ...                  ...   
2483                    1                   0                    0   
2484                    0                   0                    0   
2485                    0                   0                    0   
2486                    1                   0                    0   
2487            

## SVM Classification by using Random Search CV

In [3]:
score_measure = "precision"
kfolds = 5

param_grid = {
     'C': [0.1,1, 10, 100], 
    'gamma': [1,0.1,0.01,0.001],
    'kernel': ['poly']
}

RS_Out = SVC()
rand_search = RandomizedSearchCV(estimator = RS_Out, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestPrecisionTree = rand_search.best_estimator_

C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 16 is smaller than n_iter=500. Running 16 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 16 candidates, totalling 80 fits


C:\Users\aravi\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The best precision score is 0.9370404920282969
... with parameters: {'kernel': 'poly', 'gamma': 0.01, 'C': 0.1}


## SVM Classification by using Grid Search CV

In [5]:
score_measure = "precision"
kfolds = 5

param_grid = {
     'C': [0.1,1, 10, 100], 
     'gamma': [1,0.1,0.01,0.001],
    'kernel': ['poly']
}

GS_out = SVC()
grid_search = GridSearchCV(estimator = GS_out, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestPrecisionTree = grid_search.best_estimator_

Fitting 5 folds for each of 16 candidates, totalling 80 fits


C:\Users\aravi\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


The best precision score is 0.9370404920282969
... with parameters: {'C': 0.1, 'gamma': 0.01, 'kernel': 'poly'}


## Decision tree by using Random search cv

In [7]:
score_measure = "precision"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(1,100),  
    'min_samples_leaf': np.arange(1,100),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_leaf_nodes': np.arange(5, 100), 
    'max_depth': np.arange(1,50), 
    'criterion': ['entropy', 'gini'],
}

dtree = DecisionTreeClassifier()
rand_search = RandomizedSearchCV(estimator = dtree, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestPrecisionTree = rand_search.best_estimator_

Fitting 5 folds for each of 500 candidates, totalling 2500 fits
The best precision score is 0.8553701168468061
... with parameters: {'min_samples_split': 30, 'min_samples_leaf': 23, 'min_impurity_decrease': 0.0021, 'max_leaf_nodes': 91, 'max_depth': 35, 'criterion': 'gini'}


C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
25 fits failed out of a total of 2500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 937, in fit
    super().fit(
  File "C:\Users\aravi\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 250, in fit
    raise ValueError(
ValueError: min_samples_split must be an integer greater than 1 or a float in (

## Decision tree by using Grid search cv

In [9]:
score_measure = "precision"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(30,36),  
    'min_samples_leaf': np.arange(6,12),
    'min_impurity_decrease': np.arange(0.0048, 0.0054, 0.0001),
    'max_leaf_nodes': np.arange(162,168), 
    'max_depth': np.arange(15,21), 
    'criterion': ['entropy'],
}

Dout_GS = DecisionTreeClassifier()
grid_search = GridSearchCV(estimator = Dout_GS, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestPrecisionTree = grid_search.best_estimator_

Fitting 5 folds for each of 9072 candidates, totalling 45360 fits
The best precision score is 0.8470330066484271
... with parameters: {'criterion': 'entropy', 'max_depth': 15, 'max_leaf_nodes': 162, 'min_impurity_decrease': 0.0048, 'min_samples_leaf': 10, 'min_samples_split': 30}


# Discussion:
From the data above that the most effective approaches for the Support Vector Machine classifier are random search and grid search is 93%. After accounting for the decision tree results, the optimum grid search CV precision score is 85.3%,  Random search cv, which is 85.5%, is the approach that decision trees perform best when using. In terms of highest precision score, the svm model performs better than decision trees overall.